packages/react-components/react-combobox/src/components/Combobox/useCombobox.tsx

```diff
@@ -18,22 +20,126 @@ import type { ComboboxProps, ComboboxState } from './Combobox.types';
+
+    return searchString?.trim().toLowerCase() || '';
+  };
+
+  // set active option and selection based on typing
+  const getOptionFromInput = (inputValue: string): OptionValue | undefined => {
+    const searchString = getSearchString(inputValue);
+
+    if (searchString.length === 0) {
+      return;
+    }
+
+    const matcher = (optionValue: string) => optionValue.toLowerCase().indexOf(searchString) === 0;
+    const matches = getOptionsMatchingValue(matcher);
+
+    // return first matching option after the current active option, looping back to the top
+    if (matches.length > 1 && activeOption) {
+      const startIndex = getIndexOfId(activeOption.id);
+      const nextMatch = matches.find(option => getIndexOfId(option.id) >= startIndex);
+      return nextMatch ?? matches[0];
+    }
+
+    return matches[0] ?? undefined;
+  };
+
+  /* Handle typed input */
+
+  // reset any typed value when an option is selected
+  baseState.selectOption = (ev: SelectionEvents, option: OptionValue) => {
+    setValue(undefined);
+    selectOption(ev, option);
+  };
+
+  const onBlur = (ev: React.FocusEvent<HTMLInputElement>) => {
+    // handle selection and updating value if allowFreeform is false
+    if (!baseState.open && !allowFreeform) {
+      // select matching option, if the value fully matches
+      if (value && activeOption && getSearchString(value) === activeOption?.value.toLowerCase()) {
+        baseState.selectOption(ev, activeOption);
+      }
+
+      // reset typed value when the input loses focus while collapsed, unless allowFreeform is true
+      setValue(undefined);
+    }
+
+    setHasFocus(false);
+    triggerNativeProps.onBlur?.(ev);
+  };
+
+  baseState.setOpen = (ev, newState: boolean) => {
+    if (!newState && !allowFreeform) {
+      setValue(undefined);
+    }
+
+    setOpen(ev, newState);
+  };
+
+  // update value and active option based on input
+  const onChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
+    const inputValue = ev.target.value;
+    // update uncontrolled value
+    baseState.setValue(inputValue);
+
+    // handle updating active option based on input
+    const matchingOption = getOptionFromInput(inputValue);
+    setActiveOption(matchingOption);
+
+    // clear selection for single-select if the input value no longer matches the selection
+    if (
+      !multiselect &&
+      selectedOptions.length === 1 &&
+      (inputValue.length < 1 || selectedOptions[0].indexOf(inputValue) !== 0)
+    ) {
+      clearSelection(ev);
+    }
+
+    triggerNativeProps.onChange?.(ev);
+  };
+
+  // resolve input and listbox slot props
   const triggerShorthand = resolveShorthand(props.input, {
     required: true,
     defaultProps: {
-      onChange: () => {
-        /* Combobox does not yet support allowFreeForm */
-      },
       type: 'text',
-      value: baseState.value,
+      value: value ?? '',
       ...triggerNativeProps,
+      onBlur,
+      onChange,
+      onFocus,
     },
   });
```

Comment:
Looking at this more closely, there is a subtle potential issue for overriding the events on this slot.

By adding your onChange, onBlur, etc. functions as `defaultProps` that means they can be overridden by a custom handler on the slot: `input={{ onChange: () => { /*...*/ } }}`, and your code would never be called. Instead, you need to merge the callbacks after calling resolveShorthand. (Note the `mergeCallbacks` utility was just added in #24152, so you might need to pull latest from master).

```ts
  const trigger = resolveShorthand(props.input, {
    required: true,
    defaultProps: {
      type: 'text',
      value: value ?? '',
      ...triggerNativeProps,
    },
  });

  trigger.onFocus = mergeCallbacks(trigger.onFocus, ev => {
    // Move the onFocus code from line 46 to here, EXCEPT NOT the line that calls triggerNativeProps.onFocus
  });

  trigger.onBlur = mergeCallbacks(trigger.onBlur, ev => {
    // Move the onBlur code from lines 88-99 to here, EXCEPT NOT the line that calls triggerNativeProps.onBlur
  });

  trigger.onChange = mergeCallbacks(trigger.onChange, ev => {
    // Move the onChange code from lines 113-128 to here, EXCEPT NOT the line that calls triggerNativeProps.onChange
  });
```

This is less of an issue since `input` is the primary slot, and so most people would override the value by setting `onChange` as a prop on the Combobox instead of the `input` slot. However, it's still technically possible for them to set it on the slot. AND it's better for consistency because this is how you'd have to do it for all other slots as well.

Plus, IMO it makes the code a little cleaner, and fixes @layershifter's comment about the function names, by removing the need to give them names at all (just assign the merged function to trigger.onFocus, etc).

Reply:
good point -- and awesome that `mergeCallbacks` is available now!